### DISCUSSION TITLE
@Transactional에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#146] [Spring Data]
뭘까용?

**Comment by GIVEN53** [comment]
AOP로 어떻게 동작하는지

**Comment by 3Juhwan** [comment]
aop하면서 포함될 것 같지만 언급만 하자면 프록시 객체 내용도 추가되면 좋을 것 같습니다.

**Comment by Kimprodp** [comment]
트랜잭션은 DB에서 상태변화의 한 주기와 관련된 작업으로 모든 작업들이 성공적으로 완료되어야 작업 묶음의 결과를 적용하고, 어떤 작업에서 오류가 발생했을 때는 이전에 있던 모든 작업들이 성공적이었더라도 없었던 일처럼 완전히 되돌린다.

비즈니스로직이 트랜잭션 처리를 필요로할 때 트랜잭션 처리 코드가 비즈니스 로직과 공존한다면 코드 중복이 발생하고 비지니스 로직에 집중 또한 힘들어질 수 있다.

`@Transactional`은 이러한 문제를 해결해주는 Spring이 제공하는 어노테이션으로 `@Transactional`을 메서드 또는 클래스에 명시하게 되면 특정 메서드 또는 클래스가 제공하는 모든 메서드에 대해 내부적으로 AOP를 통해 트랜잭션 처리코드가 전 후 로 수행된다.

**[스프링의 AOP 관련 클래스]**
- Advisor: `BeanFactoryTransactionAttributeSourceAdvisor`  
- Advice: `TransactionInterceptor`  
- Pointcut: `TransactionAttributeSourcePointcut`

**[동작 순서 - 정확하지 않을 수 있음]**
1. 트랜잭션이 필요한 서비스 클래스나 메서드에 `@Transactional` 어노테이션을 추가한다.
2. 애플리케이션이 시작되면 스프링 컨텍스트가 초기화될 때 `@EnableTransactionManagement`를 통해 트랜잭션 관리가 활성화 되고, 스프링의 AOP 관련 클래스들이 빈에 등록된다.
3. `BeanFactoryTransactionAttributeSourceAdvisor`가 `TransactionInterceptor`를 사용하여 `@Transactional` 어노테이션이 있는 클래스에 트랜잭션 AOP 프록시 객체를 생성한다.
4. `@Transactional`이 적용된 메서드가 호출되면, 프록시 객체가 이 호출을 가로챈다.
5. 프록시 객체는 `TransactionInterceptor`(Transaction Advisor)를 호출하고 커넥션을 열고 닫도록 요청한다.
5. 프록시 객체는 실제 비즈니스 로직을 수행하는 메서드를 호출한다.
6. 메서드 실행이 성공하면 Transaction Advisor가 commit, rollback 등의 트랜잭션을 처리한다.

**Comment by seunghye218** [comment]
`@Transactional` 애노테이션은 AOP와 프록시 객체를 사용하여 트랜잭션을 관리합니다. `@Transactional`이 적용된 클래스의 프록시 객체를 생성하여 메서드 호출을 가로채고, 트랜잭션을 시작하고 종료하는 로직을 추가합니다. 프록시 객체는 트랜잭션 매니저를 사용하여 트랜잭션을 시작하고, 비즈니스 로직 실행 후에 트랜잭션을 커밋하거나 롤백합니다.

Spring 프레임워크의 선언적 트랜잭션 지원과 관련하여 이해해야 할 가장 중요한 개념은 이 지원이 AOP 프록시를 통해 활성화된다는 것과 트랜잭션 advice가 메타데이터(currently XML 또는 어노테이션 기반)에 의해 구동된다는 것입니다. AOP와 트랜잭션 메타데이터의 조합은 적절한 `TransactionManager` 구현과 함께 `TransactionInterceptor`를 사용하여 메서드 호출을 중심으로 트랜잭션을 구동하는 AOP 프록시를 생성합니다.

Spring 프레임워크의 `TransactionInterceptor`는 명령형 및 반응형 프로그래밍 모델을 위한 트랜잭션 관리를 제공합니다. 인터셉터는 메서드 반환 유형을 검사하여 원하는 트랜잭션 관리 방식을 감지합니다.`void`를 포함한 다른 모든 반환 유형은 명령형 트랜잭션 관리를 위해 코드 경로를 사용합니다.

명령형 트랜잭션에는 `PlatformTransactionManager`가 필요합니다. 트랜잭션은 일반적으로 `PlatformTransactionManager`가 관리하는 스레드 바인딩 트랜잭션과 함께 작동하여 현재 실행 스레드 내의 모든 데이터 액세스 작업에 트랜잭션을 노출합니다. 메서드 내에서 새로 시작된 스레드에는 전파되지 않습니다.

다음 이미지는 트랜잭션 프록시에서 메서드를 호출하는 개념도를 보여줍니다.
![image](https://github.com/2024-woowacourse-study/level-interview/assets/93831492/9bf72f7c-1586-4db3-b8dc-2cba0664ba02)

다음은 Spring 프레임워크에서 트랜잭션 처리 중 발생한 UnsupportedOperationException 예외 상황입니다.
```java
<!-- the Spring container is starting up... -->
[AspectJInvocationContextExposingAdvisorAutoProxyCreator] - Creating implicit proxy for bean 'fooService' with 0 common interceptors and 1 specific interceptors

<!-- the DefaultFooService is actually proxied -->
[JdkDynamicAopProxy] - Creating JDK dynamic proxy for [x.y.service.DefaultFooService]

<!-- ... the insertFoo(..) method is now being invoked on the proxy -->
[TransactionInterceptor] - Getting transaction for x.y.service.FooService.insertFoo

<!-- the transactional advice kicks in here... -->
[DataSourceTransactionManager] - Creating new transaction with name [x.y.service.FooService.insertFoo]
[DataSourceTransactionManager] - Acquired Connection [org.apache.commons.dbcp.PoolableConnection@a53de4] for JDBC transaction

<!-- the insertFoo(..) method from DefaultFooService throws an exception... -->
[RuleBasedTransactionAttribute] - Applying rules to determine whether transaction should rollback on java.lang.UnsupportedOperationException
[TransactionInterceptor] - Invoking rollback for transaction on x.y.service.FooService.insertFoo due to throwable [java.lang.UnsupportedOperationException]

<!-- and the transaction is rolled back (by default, RuntimeException instances cause rollback) -->
[DataSourceTransactionManager] - Rolling back JDBC transaction on Connection [org.apache.commons.dbcp.PoolableConnection@a53de4]
[DataSourceTransactionManager] - Releasing JDBC Connection after transaction
[DataSourceUtils] - Returning JDBC Connection to DataSource

Exception in thread "main" java.lang.UnsupportedOperationException at x.y.service.DefaultFooService.insertFoo(DefaultFooService.java:14)
<!-- AOP infrastructure stack trace elements removed for clarity -->
at $Proxy0.insertFoo(Unknown Source)
at Boot.main(Boot.java:11)
```
